### DISCUSSION TITLE
qusetion to vips_crop ()

### DISCUSSION THREAD

**Discussion by rebornfz** [libvips/libvips#2271] [Q&A] [answered]
How can I intercept a circle or a polygon？
Like this：
![3cd95070-bf8b-11eb-b85b-0a580a431449](https://user-images.githubusercontent.com/27042311/119956840-1585e300-bfd4-11eb-8d4b-ccd90223ee27.png)
circle 
![66](https://user-images.githubusercontent.com/27042311/119956877-1d458780-bfd4-11eb-92bf-943b50284268.PNG)
polygon
![获](https://user-images.githubusercontent.com/27042311/119956891-2171a500-bfd4-11eb-81e1-b6002acc6d52.PNG)

Or How can I draw a polygon whose alpha channel is transparent

**Comment by jcupitt** [comment]
Hi again,

It's probably easiest with SVG -- the libvips SVG loader is very fast and can make huge images while only needing a little memory. For example:

```python
#!/usr/bin/python3

import sys
import pyvips

image = pyvips.Image.new_from_file(sys.argv[1])

cx = image.width / 2
cy = image.height / 2
r = min(cx, cy)
mask = pyvips.Image.svgload_buffer(bytes(f"""
<svg viewBox="0 0 {image.width} {image.height}">
    <circle cx="{cx}" cy="{cy}" r="{r}" fill="#000" />
</svg>
""", "utf8"))

# svgload always makes RGBA ... we want the alpha
alpha = mask[3]

image.bandjoin(alpha).write_to_file(sys.argv[2])
```

**Reply by rebornfz** [reply]
@jcupitt 
But I use png, Can png do the same？

**Reply by jcupitt** [reply]
Sure, drop the alpha off the input image (if any -- it's optional in PNG) and swap in the alpha from the SVG.

**Reply by rebornfz** [reply]
Can I directly capture a circular area from the image?

**Reply by jcupitt** [reply]
Yes, that's the alpha.

**Reply by jcupitt** [reply]
If you need to combine an alpha on the input image with the circular mask alpha, I would use `composite`. One of the modes does what you want.

**Reply by rebornfz** [reply]
@jcupitt  How can I use this in c++ api？

**Reply by jcupitt** [reply]
glib has `g_strdup_printf` which is handy for stuff like this, and is included in libvips:

```C++
char *svg = g_strdup_printf("
<svg viewBox=\"0 0 %d %d\">
    <circle cx=\"%g" cy="%g\" r=\"%g\" fill=\"#000\" />
</svg>
",
    image.width, image.height, cx, cy, r);
mask = VImage::svgload_bvuffer(svg);
g_free(svg);
...
```

Or you can use`format` if you're lucky enough to have C++20.

**Reply by rebornfz** [reply]
@jcupitt  Thanks for your reply. I get this error when I try this method.

error: no matching function for call to 'vips::VImage::svgload_buffer(char*&)'
    vips::VImage mask = vips::VImage::svgload_buffer(svg);
/usr/include/vips/VImage8.h:5313:15: note: candidate: 'static vips::VImage vips::VImage::svgload_buffer(VipsBlob*, vips::VOption*)'
 static VImage svgload_buffer( VipsBlob *buffer, VOption *options = 0 );
               ^~~~~~~~~~~~~~
/usr/include/vips/VImage8.h:5313:15: note:   no known conversion for argument 1 from 'char*' to 'VipsBlob*' {aka '_VipsBlob*'}

**Reply by rebornfz** [reply]
@jcupitt  
Hello, Is the above question a problem with the method I used?

**Reply by rebornfz** [reply]
@jcupitt 
There seems to be no such usage in the c api.
![捕获](https://user-images.githubusercontent.com/27042311/120277975-4f622c80-c2e7-11eb-96fe-0d73e9961765.PNG)
![捕获1](https://user-images.githubusercontent.com/27042311/120277984-512bf000-c2e7-11eb-8aca-301195281eb7.PNG)

**Comment by jcupitt** [comment, marked as answer]
Ooop, you're right, sorry.

Yes, `svgload_buffer` takes a blob, not a string, since you can give it binary data (sgz).

Try:

```C++
VipsBlob *blob = vips_blob_new( NULL, str.c_str(), str.length() );
VImage im = VImage::svgload_buffer(blob);
vips_area_unref( VIPS_AREA( blob ) );
```

**Reply by rebornfz** [reply]
@jcupitt 
Thank you! It worked well. My code:
```
char *svg = g_strdup_printf("<svg viewBox=\"0 0 %d %d\"><circle cx=\"%g\" cy=\"%g\" r=\"%g\" fill=\"#000\" /></svg>", c_width, c_height, cx, cy, r);
std::string str = svg;
VipsBlob *blob = vips_blob_new( NULL, str.c_str(), str.length() );
vips::VImage mask = vips::VImage::svgload_buffer(blob);
g_free(svg);
vips_area_unref( VIPS_AREA( blob ) );
image_input = image_input.composite2(mask, VIPS_BLEND_MODE_DEST_IN);
```

**Reply by jcupitt** [reply]
Thinking about it, you could also use:

```C++
char *svg = g_strdup_printf("<svg viewBox=\"0 0 %d %d\"><circle cx=\"%g\" cy=\"%g\" r=\"%g\" fill=\"#000\" /></svg>", c_width, c_height, cx, cy, r);
vips::VImage mask = vips::VImage::new_from_buffer(std::string(str), "");
```

A little neater.
